### PR TITLE
Use Agent for Capi Merchandising Component Content

### PR DESCRIPTION
**Pull request by kelvin-chappell**
This is to avoid making unnecessary calls on Capi.

### CODE REVIEW

File: commercial/app/model/commercial/CapiAgent.scala [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      shortUrlAgent alter (_ ++ newContents)
+    }
+
+    val eventualNewCache = if (urlsNotInCache.isEmpty) {
+      Future.successful(cache)
+    } else {
+      Lookup.contentByShortUrls(shortUrlIds) flatMap {
+        addToCache
+      } recoverWith {
+        case NonFatal(e) => Future.successful(cache)
```
Comment (janua) [inline commercial/app/model/commercial/CapiAgent.scala]: Do you want any metrics or logging here?

File: commercial/app/model/commercial/CapiAgent.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      val slashPrefixed = if (url startsWith "/") url else s"/$url"
+      val endTrimmed = slashPrefixed.take(slashPrefixed.indexOf("/", 3))
+      endTrimmed
+    }
+
+    val cache = shortUrlAgent.get()
+
+    val urlsNotInCache = shortUrlIds filterNot cache.contains
+
+    def addToCache(contents: Seq[Content]): Future[Map[String, Option[Content]]] = {
```
Comment (janua) [inline commercial/app/model/commercial/CapiAgent.scala]: Why not have this inside the `object`? It's a matter of taste but I didn't spot it straight away that all these statements were inside a `def`!

**Comment by janua** [comment]
Maybe consider lazy vals?

:+1:

**Comment by kelvin-chappell** [comment]
@janua I've considered and changed `cache` to a lazy val

File: commercial/app/model/commercial/CapiAgent.scala (hunk 1 above)
Comment (kelvin-chappell) [inline commercial/app/model/commercial/CapiAgent.scala]: I think I'll avoid them initially.  I find too much logging makes the code more difficult to read and just adds noise.  I can't imagine there's a metric anyone will look at.  They'll just want to know why the endpoint isn't working.

Comment (janua) [inline commercial/app/model/commercial/CapiAgent.scala]: But you are basically exception swallowing, so if there is a problem here you will never know.

If you really don't care about knowing though then I guess it's fine :smile:

**Comment by janua** [comment]
:+1:

File: commercial/app/model/commercial/CapiAgent.scala (hunk 1 above)
Comment (kelvin-chappell) [inline commercial/app/model/commercial/CapiAgent.scala]: I take your point.  I should be logging that exception.  There ought to be a better highlight in the UI of which line you're commenting on.

**Pull request merged**